File: cucumber/src/main/java/com/saucedemo/utils/SlackUtils.java
package com.saucedemo.utils;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;

import java.util.Map;

/**
 * SlackUtils: Dispatches real-time test execution summaries to a Slack channel.
 * Webhook URL is read from the SLACK_WEBHOOK_URL environment variable / secret.
 */
public class SlackUtils {

    public static void sendResult(String message) {
        String webhookUrl = System.getenv("SLACK_WEBHOOK_URL");

        if (webhookUrl == null || webhookUrl.isEmpty()) {
            System.out.println("[WARN] SLACK_WEBHOOK_URL not set. Skipping Slack notification.");
            return;
        }

        try {
            String jsonPayload = new ObjectMapper()
                    .writeValueAsString(Map.of("text", message));

            RestAssured.given()
                    .contentType(ContentType.JSON)
                    .body(jsonPayload)
                    .post(webhookUrl)
                    .then()
                    .statusCode(200);

            System.out.println("[INFO] Slack notification sent successfully.");
        } catch (Exception e) {
            System.err.println("[ERROR] Slack notification failed: " + e.getMessage());
        }
    }
}

File: cucumber_python/utils/slack_utils.py
"""
slack_utils.py — Send suite result notifications to a Slack webhook.

Fail-safe: if SLACK_WEBHOOK_URL is absent, logs a warning and returns.
Mirrors SlackUtils.java.
"""

import os
import requests


def send_result(message: str) -> None:
    """Post a message to the configured Slack webhook.

    Args:
        message: The text to post (supports Slack mrkdwn formatting).
    """
    webhook_url = os.getenv("SLACK_WEBHOOK_URL", "")
    if not webhook_url:
        print("[WARN] SLACK_WEBHOOK_URL not set. Skipping Slack notification.")
        return

    try:
        response = requests.post(
            webhook_url,
            json={"text": message},
            headers={"Content-Type": "application/json"},
            timeout=10,
        )
        if response.status_code == 200:
            print("[INFO] Slack notification sent.")
        else:
            print(f"[WARN] Slack notification returned status: {response.status_code}")
    except Exception as exc:
        print(f"[ERROR] Slack notification failed: {exc}")
